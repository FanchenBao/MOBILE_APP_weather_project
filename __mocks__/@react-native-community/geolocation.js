// mock @react-native-community/geolocation
// This module is not directly used in the project, but used by
// react-native-geolocation-service

export default {
  addListener: jest.fn(),
  getCurrentPosition: jest.fn(),
  removeListeners: jest.fn(),
  requestAuthorization: jest.fn(),
  setConfiguration: jest.fn(),
  startObserving: jest.fn(),
  stopObserving: jest.fn(),
};
