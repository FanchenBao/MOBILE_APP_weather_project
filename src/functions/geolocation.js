import {PermissionsAndroid} from 'react-native';
import Geolocation from 'react-native-geolocation-service';
import {getFineLocationPermission} from './get_user_permission.js';

/**
 * Obtain geo location from the device.
 *
 * This function first checks whether permission has been granted for obtaining
 * geo location. If yes, it proceeds to get the geo location. If not, it asks
 * user for the permission.
 *
 * @param {Function} successCallback Callback when geo location has been
 * successfully obtained.
 * @param {Function} errorCallback Callback when there is an error in getting
 * geo location.
 * @param {Function} denyCallback Callback when user explicitly denies geo
 * location access.
 */
async function getGeolocation(successCallback, errorCallback, denyCallback) {
  try {
    let hasLocationPermission = await PermissionsAndroid.check(
      PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
    );
    if (hasLocationPermission) {
      Geolocation.getCurrentPosition(successCallback, errorCallback, {
        enableHighAccuracy: true,
        timeout: 15000,
        maximumAge: 10000,
      });
    } else {
      console.log('Not allowed to access fine location. Ask for permission');
      let userDecision = await getFineLocationPermission();
      if (userDecision === PermissionsAndroid.RESULTS.GRANTED) {
        console.log('Geolocation permission granted');
        await getGeolocation(successCallback, errorCallback, denyCallback);
      } else {
        console.log(`Geolocation denied. User decision: ${userDecision}`);
        denyCallback();
      }
    }
  } catch (error) {
    console.error(error);
  }
}

export {getGeolocation};
