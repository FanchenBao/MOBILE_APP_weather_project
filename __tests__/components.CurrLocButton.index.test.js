import {CurrLocButton} from '../src/components/CurrLocButton/index.js';
import React from 'react';
import renderer from 'react-test-renderer';
import {getGeolocation} from '../src/functions/geolocation.js';
import {TouchableHighlight} from 'react-native';

let component;
let touchableHighlight;
let mockConsoleError;
let _setStateForecast;
let _setToWait;
// set up mocking for mockConsoleError. Also create the CurrLocButton component.
beforeEach(() => {
  // mock the function passed as props to CurrLocButton
  _setStateForecast = jest.fn(props => props);
  _setToWait = jest.fn(() => null);

  mockConsoleError = jest.spyOn(global.console, 'error').mockImplementation();
  // create a CurrLocButton component
  component = renderer.create(
    <CurrLocButton
      setStateForecast={_setStateForecast}
      setToWait={_setToWait}
      hasInternet={true}
    />,
  );
  // selected TouchableHighLight component
  touchableHighlight = component.root.findByType(TouchableHighlight);
});

// clean up console function mock after each test.
afterEach(() => {
  mockConsoleError.mockRestore();
});

// Mock app's own module
jest.mock('../src/functions/geolocation.js');
getGeolocation
  .mockImplementationOnce(
    async (successCallback, errorCallback, denyCallback) => {
      successCallback({coords: {latitude: 10, longitude: 20}}); // test 2
    },
  )
  .mockImplementationOnce(
    async (successCallback, errorCallback, denyCallback) => {
      denyCallback(); // test 3
    },
  )
  .mockImplementationOnce(
    async (successCallback, errorCallback, denyCallback) => {
      errorCallback({code: 1, message: 'foo'}); // test 4
    },
  )
  .mockRejectedValue('bar'); // test 5

// test 1
test('Snapshot on CurrLocButton', () => {
  expect(component.toJSON()).toMatchSnapshot();
});

// test 2
test('getGeolocation successCalback', async () => {
  await touchableHighlight.props.onPress();
  expect(_setStateForecast).toHaveBeenCalledTimes(1);
  expect(_setStateForecast).toHaveBeenCalledWith({
    type: 'coord',
    value: {
      lat: 10,
      lon: 20,
    },
  });
  expect(_setToWait).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).not.toHaveBeenCalled();
});

// test 3
test('getGeolocation denyCallback', async () => {
  await touchableHighlight.props.onPress();
  expect(_setStateForecast).toHaveBeenCalledTimes(1);
  expect(_setStateForecast).toHaveBeenCalledWith({
    type: 'error',
    value: 'Obtain current location NOT allowed',
  });
  expect(_setToWait).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).not.toHaveBeenCalled();
});

// test 4
test('getGeolocation errorCallback', async () => {
  await touchableHighlight.props.onPress();
  expect(_setStateForecast).toHaveBeenCalledTimes(1);
  expect(_setStateForecast).toHaveBeenCalledWith({
    type: 'error',
    value: 'Obtain current location failed',
  });
  expect(_setToWait).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).toHaveBeenCalledWith(1, 'foo');
});

test('getGeolocation unexpected error', async () => {
  await touchableHighlight.props.onPress();
  expect(_setStateForecast).not.toHaveBeenCalled();
  expect(_setToWait).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).toHaveBeenCalledTimes(1);
  expect(mockConsoleError).toHaveBeenCalledWith('bar');
});
