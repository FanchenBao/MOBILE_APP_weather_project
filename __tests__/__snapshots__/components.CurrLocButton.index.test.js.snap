// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot on CurrLocButton 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 10,
      "flex": 0.5,
      "justifyContent": "center",
      "marginHorizontal": 100,
      "marginVertical": 5,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 18,
      }
    }
  >
    Use Current Location
  </Text>
</View>
`;
